### ISSUE TITLE
Cryptical error message when referer not provided

### ISSUE THREAD

**Issue by warpech** [Palindrom/Palindrom#49]
When `referer` is not provided and Puppet is configured to use WebSocket protocol, the error message is too cryptical.

See the screenshot:

![cryptical](https://cloud.githubusercontent.com/assets/566463/7180549/e5cd457c-e442-11e4-887c-048dd066bbf5.png)

**Comment by warpech** [comment]
Splitting the discussion from #58 (which is about the acutal JavaScript error).

This issue (#49) is about **How a Starcounter app developer should get notified by Puppet that he forgot to create a session.**

By now we have contradicting expectations - we want to leave it up to the developer, but the default option allows to shoot yourself in the foot with unclear error.

Puppet protocol does not actually require server to set `Location / X-Location / Cookie: Location` header. It is server's privilege of doing it, but in case the server does not do it, the client will just use the page URL for subsequent requests.

After #58 is fixed, the developer will see no error in such case. I think the correct error behavior should be that the server rejects `application/json-patch+json` request or WebSocket upgrade if there is no session.

That should solve it. I will investigate what is the current behavior and assign the tasks.

**Comment by warpech** [comment]
After recent fix in #58, now we have a more meaningful cavalcade of errors:

![image](https://cloud.githubusercontent.com/assets/566463/8461451/f69d8bac-202c-11e5-981e-febd2848a71c.png)

These errors come because client tries to upgrate to WebSocket when there is no session:

```
WebSocket connection to 'ws://localhost:8080/invoicedemo/new-invoice' failed: Error during WebSocket handshake: Unexpected response code: 200
Uncaught Error: WebSocket connection could not be made.
  Could not connect to: ws://localhost:8080/invoicedemo/new-invoicePuppetNetworkChannel.webSocketUpgrade.that._ws.onerror
WebSocket connection closed. Status code: 1006.
```

These errors come because client falls back to HTTP PATCH when there is no session:

```
PATCH http://localhost:8080/invoicedemo/new-invoice 404 (Not Found)
Uncaught Error: PuppetJs JSON response error. Server responded with error 404 Not Found
  URI not found: PATCH /invoicedemo/new-invoice 
```

This is better than before, but still not perfect. Why WebSocket upgrade responds with 200 OK and HTTP PATCH responds with 404 Not Found? It needs to be discussed and fixed on the server side.

**Comment by warpech** [comment]
The issue was reported today again by @niklasbjorkman. 

When you forget to set `page.Session` in your C#, this is the error that you get:

![2016-04-14 12_38_44-client-side using polymer](https://cloud.githubusercontent.com/assets/566463/14525536/33c95260-023e-11e6-97f5-6728754122d7.png)

This is super confusing. I will think of a solution.

**Comment by warpech** [comment]
@alshakero how is this presented now, since `remoteUrl` is required?

**Comment by alshakero** [comment]
@warpech I've never seen this issue. It's very important. 

I'm still not happy with the way Palindrom deals with URLs. The current scenario is:

1. Remote URL is given and is **only** used for HTTP initial request. Even if it was the correct session URL, WebSocket will not upgrade until the HTTP response is received. 
2. The response comes with a `Location` header that Palindrom uses to start a WebSocket connection. If the response didn't have that header, the WebSocket connection will be to the same URL (with `ws` scheme). 

And this makes remoteURL useless in most cases, the needed information is within `Location` header anyways. And if the information was not there, we are still waiting for HTTP request to finish.

I think a video conversation would do better. My thoughts aren't clear on the issue.

**Comment by warpech** [comment]
> And this makes remoteURL useless in most cases, the needed information is within Location header anyways. And if the information was not there, we are still waiting for HTTP request to finish.

Remote URL is needed to (1) obtain the state of the session `application/json` as well as to send (2) HTTP Patch `application/json-patch+json` changes to it. WebSocket is just an upgrade to the part (2).

So I would not call it "mostly useless", I would call it "fundamental". Let's have a chat on HO if you please.

**Comment by alshakero** [comment]
@warpech, Of course, Remote URL is important. I just mean that in effect we are only using the one in `Location` header.

**Comment by alshakero** [comment]
@warpech I am ready for HO anytime today.

**Comment by warpech** [comment]
We decided on HO that this issue is no longer relevant since now ~`referer`~ `removeUrl` parameter is required.
